beacon/src/region.rs

```diff
@@ -0,0 +1,203 @@
+    type Error = Error;
+    fn try_from(value: GatewayRegionParamsRespV1) -> Result<Self> {
+        let region = Region::from_i32(value.region)?;
+        let params = value
+            .params
+            .ok_or_else(Error::no_region_params)?
+            .region_params;
+        Ok(Self {
+            gain: Decimal::new(value.gain as i64, 1),
+            params,
+            region,
+        })
+    }
+}
+// This is the currently minimimum conducted power supported by the semtech
+// packet forwarder
+lazy_static::lazy_static! {
+    static ref MIN_CONDUCTED_POWER: Decimal = Decimal::new(120, 1);
+}
+
+impl RegionParams {
+    pub fn max_eirp(&self) -> Result<Decimal> {
+        self.params
+            .iter()
+            .max_by_key(|p| p.max_eirp)
+            .ok_or_else(Error::no_region_params)
+            .map(|v| Decimal::new(v.max_eirp as i64, 1))
+    }
+
+    pub fn max_conducted_power(&self) -> Result<u32> {
+        self.max_eirp().and_then(|max_eirp| {
+            (max_eirp - self.gain)
+                .trunc()
+                .to_u32()
+                .ok_or_else(Error::invalid_conducted_power)
+        })
+    }
+
+    pub fn min_conducted_power(&self) -> Result<u32> {
+        (*MIN_CONDUCTED_POWER)
+            .trunc()
+            .to_u32()
+            .ok_or_else(Error::invalid_conducted_power)
+    }
+
+    /// Selects the best spreading factor for the given packet size
+    pub fn select_spreading(&self, packet_size: usize) -> Result<RegionSpreading> {
+        // The spreading does not change per channel frequency So just get one
+        // and do selection depending on max_packet_size
+        self.params
+            .first()
+            // ensure region params themselves are available
+            .ok_or_else(Error::no_region_params)
+            // Ensure spreading parameters are available
+            .and_then(|params| {
+                params
+                    .spreading
+                    .as_ref()
+                    .ok_or_else(Error::no_region_spreading)
+            })
+            // Find the first tagged spreading that can manage the given
+            // packet size and extract the region spreading
+            .and_then(|spreading| {
+                spreading
+                    .tagged_spreading
+                    .iter()
+                    .find(|tagged_spreading| {
+                        tagged_spreading.max_packet_size as usize >= packet_size
+                    })
+                    .map(|tagged_spreading| tagged_spreading.region_spreading)
+                    .ok_or_else(|| Error::no_region_spreading_for_size(packet_size))
+            })
+            // Convert to RegionSpreading
+            .and_then(|region_spreading| {
+                RegionSpreading::from_i32(region_spreading)
+                    .ok_or_else(|| Error::unsupported_region_spreading(region_spreading))
+            })
+    }
+
+    pub fn bandwitdh(&self) -> Result<u32> {
```

Comment:
this is why it compiled @ke6jjj ; it's a _consistent_ typo 😁

Reply:
No chance to fix, eh? it's in a different crate?